File: internal/workflow/validate.go
package workflow

import (
	"fmt"
	"regexp"
	"slices"
	"strings"

	"github.com/qoggy/conduct/internal/apperror"
	"github.com/qoggy/conduct/internal/engine"
)

// nodeIDPattern 限定 agent 节点 id：首字符字母 / 下划线，其余字母 / 数字 / 连字符 / 下划线，总长 1–64。
// 注意与工作流名规则不同（工作流名见 name.go，可含点、可数字开头）。
var nodeIDPattern = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_-]{0,63}$`)

// templateVariablePattern 匹配模板里的变量引用 {{key}}，允许前置反斜杠转义 \{{key}}。
// 与运行时 render 阶段保持同一套语法。
var templateVariablePattern = regexp.MustCompile(`\\?\{\{([a-zA-Z_][\w.-]*)\}\}`)

// knownSystemVariables 是模板里允许的 {{sys.*}} 系统变量集。
var knownSystemVariables = map[string]bool{
	"sys.userPrompt": true,
	"sys.cwd":        true,
	"sys.runId":      true,
}

// IsValidNodeID 报告 id 是否符合 agent 节点命名规则（不含保留名判断，供 CLI node add 在建节点前速判 id 格式）。
func IsValidNodeID(id string) bool { return nodeIDPattern.MatchString(id) }

// Problem 是一条字段级校验错误：Path 为稳定字段路径，Code/Params 为可由 CLI/UI 分别本地化的机器信息。
type Problem = apperror.Problem

// ValidateStructured 对一份定义主体执行落盘校验，逐条收集字段级错误后一并返回（承 spec〈落盘校验规则〉：
// 恰好一个 START/END + ≥1 agent、保留名、标记节点必空、边合法、无环、单源单汇无悬空、模板引用祖先）。
// 不修改 def。返回空切片表示校验通过。Validate 把问题包装成共享结构化错误。
func ValidateStructured(def *Definition) []Problem {
	if len(def.Nodes) == 0 {
		return []Problem{{Path: "nodes", Code: apperror.CodeNodesRequired}}
	}

	var problems []Problem

	// 预扫描：node id → 首次位置（供唯一性与引用校验），并数 START / END / agent。
	indexByID := make(map[string]int, len(def.Nodes))
	startCount, endCount, agentCount := 0, 0, 0
	for position, node := range def.Nodes {
		switch {
		case node.IsStart():
			startCount++
		case node.IsEnd():
			endCount++
		default:
			agentCount++
		}
		if node.ID == "" {
			continue // 空 id 的必填错误在下方 agent 主循环报出
		}
		if _, duplicate := indexByID[node.ID]; duplicate {
			problems = append(problems, problem(fmt.Sprintf("nodes[%d].id", position), apperror.CodeDuplicateNodeID, "id", node.ID))
			continue
		}
		indexByID[node.ID] = position
	}
	validNodeID := func(id string) bool { _, ok := indexByID[id]; return ok }

	// —— 节点集：恰好一个 START、一个 END，另有 ≥1 agent 节点 ——
	if startCount != 1 {
		problems = append(problems, problem("nodes", apperror.CodeStartNodeCount, "count", startCount))
	}
	if endCount != 1 {
		problems = append(problems, problem("nodes", apperror.CodeEndNodeCount, "count", endCount))
	}
	if agentCount == 0 {
		problems = append(problems, Problem{Path: "nodes", Code: apperror.CodeAgentNodeRequired})
	}

	// —— 逐节点：标记节点必空 / agent 节点必填与能力表 ——
	for position, node := range def.Nodes {
		path := fmt.Sprintf("nodes[%d]", position)
		if node.IsMarker() {
			problems = append(problems, validateMarkerNode(path, node)...)
			continue
		}
		problems = append(problems, validateAgentNode(path, node)...)
	}

	// —— 边：引用存在、自环、重复、指向 START、源自 END、START→END 直连 ——
	problems = append(problems, validateEdges(def, validNodeID)...)

	// —— 单源单汇 / 无悬空：每个 agent 节点 ≥1 入边、≥1 出边 ——
	problems = append(problems, validateDegrees(def)...)

	// —— 无环 ——
	cycle := DetectCycle(def)
	if cycle != nil {
		problems = append(problems, problem("edges", apperror.CodeCycleDetected, "cycle", strings.Join(cycle, "→")))
	}

	// —— 模板引用祖先 —— 仅在无环时「祖先」有定义，故 DetectCycle 通过后再算。
	if cycle == nil {
		problems = append(problems, validateTemplateAncestry(def, validNodeID)...)
	}

	return problems
}

// Validate 是 ValidateStructured 的字符串化包装：把每条 Problem 拼成 "<path>: <message>"、换行连接返回。
// 各调用方（CLI / store 载入兜底）按整串比对 / 子串断言。无错误返回 nil。
func Validate(def *Definition) error {
	problems := ValidateStructured(def)
	if len(problems) == 0 {
		return nil
	}
	return apperror.Validation(problems)
}

// validateMarkerNode 校验 START / END 标记节点：engine / promptTemplate / engineConfig / displayName 必空
// （标记节点不承载配置与展示名，UI 直接渲染其 id）。
func validateMarkerNode(path string, node Node) []Problem {
	var problems []Problem
	if node.DisplayName != "" {
		problems = append(problems, markerFieldProblem(path+".displayName", node.ID, "displayName"))
	}
	if node.Engine != "" {
		problems = append(problems, markerFieldProblem(path+".engine", node.ID, "engine"))
	}
	if node.PromptTemplate != "" {
		problems = append(problems, markerFieldProblem(path+".promptTemplate", node.ID, "promptTemplate"))
	}
	if node.EngineConfig != nil {
		problems = append(problems, markerFieldProblem(path+".engineConfig", node.ID, "engineConfig"))
	}
	return problems
}

// validateAgentNode 校验 agent 节点的必填字段与格式、engine + engineConfig 能力表。
// 模板变量引用（须皆祖先）在 validateTemplateAncestry 里另行校验（依赖无环后的祖先集）。
func validateAgentNode(path string, node Node) []Problem {
	var problems []Problem
	if node.ID == "" {
		problems = append(problems, requiredProblem(path+".id", "id"))
	} else if !nodeIDPattern.MatchString(node.ID) {
		problems = append(problems, problem(path+".id", apperror.CodeInvalidNodeID, "id", node.ID))
	}
	if node.DisplayName == "" {
		problems = append(problems, requiredProblem(path+".displayName", "displayName"))
	}
	if node.PromptTemplate == "" {
		problems = append(problems, requiredProblem(path+".promptTemplate", "promptTemplate"))
	}
	problems = append(problems, validateEngine(path, node.Engine, node.EngineConfig)...)
	return problems
}

// validateEdges 校验每条边：端点存在（可含 START/END）、禁止自环 / 重复边 / 边指向 START / 边源自 END /
// START→END 直连。
func validateEdges(def *Definition, validNodeID func(string) bool) []Problem {
	var problems []Problem
	seen := make(map[string]bool, len(def.Edges))
	for i, edge := range def.Edges {
		path := fmt.Sprintf("edges[%d]", i)
		if edge.From == "" || edge.To == "" {
			problems = append(problems, Problem{Path: path, Code: apperror.CodeEdgeEndpointsRequired})
			continue
		}
		if !validNodeID(edge.From) {
			problems = append(problems, problem(path, apperror.CodeEdgeFromNodeNotFound, "id", edge.From))
		}
		if !validNodeID(edge.To) {
			problems = append(problems, problem(path, apperror.CodeEdgeToNodeNotFound, "id", edge.To))
		}
		if edge.From == edge.To {
			problems = append(problems, problem(path, apperror.CodeSelfEdge, "from", edge.From, "to", edge.To))
		}
		if edge.From == NodeIDStart && edge.To == NodeIDEnd {
			problems = append(problems, Problem{Path: path, Code: apperror.CodeStartEndDirectEdge})
		} else {
			if edge.To == NodeIDStart {
				problems = append(problems, Problem{Path: path, Code: apperror.CodeEdgeToStart})
			}
			if edge.From == NodeIDEnd {
				problems = append(problems, Problem{Path: path, Code: apperror.CodeEdgeFromEnd})
			}
		}
		key := edge.From + "\x00" + edge.To
		if seen[key] {
			problems = append(problems, problem(path, apperror.CodeDuplicateEdge, "from", edge.From, "to", edge.To))
		}
		seen[key] = true
	}
	return problems
}

// validateDegrees 校验每个 agent 节点 ≥1 入边、≥1 出边（单源单汇的落地：入度 0 只允许 START、出度 0 只允许
// END）。标记节点的度约束由 validateEdges 兜住（禁止指向 START / 源自 END）。
func validateDegrees(def *Definition) []Problem {
	inDegree := make(map[string]int, len(def.Nodes))
	outDegree := make(map[string]int, len(def.Nodes))
	for _, edge := range def.Edges {
		outDegree[edge.From]++
		inDegree[edge.To]++
	}
	var problems []Problem
	for position, node := range def.Nodes {
		if !node.IsAgent() {
			continue
		}
		path := fmt.Sprintf("nodes[%d]", position)
		if inDegree[node.ID] == 0 {
			problems = append(problems, problem(path, apperror.CodeNodeMissingIncomingEdge, "id", node.ID))
		}
		if outDegree[node.ID] == 0 {
			problems = append(problems, problem(path, apperror.CodeNodeMissingOutgoingEdge, "id", node.ID))
		}
	}
	return problems
}

// validateTemplateAncestry 校验每个 agent 节点 promptTemplate 里非转义的 {{X}}：
// {{sys.*}} 限已知系统变量；{{START}} / {{END}} 禁止（标记节点无产物）；其余须引用本节点的上游祖先 agent
// 节点（存在但非祖先、或根本不存在，均拒绝）。假定图无环（由调用方在 DetectCycle 通过后再调）。
func validateTemplateAncestry(def *Definition, validNodeID func(string) bool) []Problem {
	var problems []Problem
	for position, node := range def.Nodes {
		if !node.IsAgent() {
			continue
		}
		ancestors := Ancestors(def, node.ID)
		path := fmt.Sprintf("nodes[%d].promptTemplate", position)
		for _, match := range templateVariablePattern.FindAllStringSubmatch(node.PromptTemplate, -1) {
			full, key := match[0], match[1]
			if strings.HasPrefix(full, "\\") {
				continue // 转义 \{{x}} → 字面量，不校验
			}
			if strings.HasPrefix(key, "sys.") {
				if !knownSystemVariables[key] {
					problems = append(problems, problem(path, apperror.CodeUnknownSystemVariable, "key", key))
				}
				continue
			}
			if key == NodeIDStart || key == NodeIDEnd {
				problems = append(problems, problem(path, apperror.CodeMarkerNodeReference, "id", key))
				continue
			}
			if ancestors[key] {
				continue
			}
			if validNodeID(key) {
				problems = append(problems, problem(path, apperror.CodeNonAncestorNodeReference, "id", key))
			} else {
				problems = append(problems, problem(path, apperror.CodeNodeReferenceNotFound, "id", key))
			}
		}
	}
	return problems
}

// validateEngine 校验 engine 合法性及其 engineConfig（判别联合，依 engine 包能力表）。
func validateEngine(path, engineName string, config *EngineConfig) []Problem {
	if engineName == "" {
		return []Problem{requiredProblem(path+".engine", "engine")}
	}
	if !engine.Exists(engineName) {
		return []Problem{problem(path+".engine", apperror.CodeUnknownEngine,
			"engine", engineName, "available", strings.Join(engine.RegisteredNames(), ", "))}
	}
	if config == nil {
		return nil
	}

	var problems []Problem
	descriptor, _ := engine.Describe(engineName)
	capability := descriptor.Capability

	if config.Model != "" && !capability.AllowsModel {
		problems = append(problems, problem(path+".engineConfig.model", apperror.CodeEngineModelNotAllowed, "engine", engineName))
	}
	if config.Effort != "" {
		if !capability.AllowsEffort {
			problems = append(problems, effortFieldProblem(path+".engineConfig.effort", engineName))
		} else if !slices.Contains(capability.EffortValues, config.Effort) {
			problems = append(problems, effortValueProblem(path+".engineConfig.effort", engineName, config.Effort, capability.EffortValues))
		}
	}
	return problems
}

func problem(path string, code apperror.Code, pairs ...any) Problem {
	params := make(apperror.Params, len(pairs)/2)
	for index := 0; index+1 < len(pairs); index += 2 {
		params[pairs[index].(string)] = pairs[index+1]
	}
	if len(params) == 0 {
		params = nil
	}
	return Problem{Path: path, Code: code, Params: params}
}

func markerFieldProblem(path, id, field string) Problem {
	return problem(path, apperror.CodeMarkerFieldNotEmpty, "id", id, "field", field)
}

func requiredProblem(path, field string) Problem {
	return problem(path, apperror.CodeRequiredField, "field", field)
}

func effortFieldProblem(path, engineName string) Problem {
	return problem(path, apperror.CodeEngineEffortFieldNotAllowed, "engine", engineName)
}

func effortValueProblem(path, engineName, value string, allowed []string) Problem {
	return problem(path, apperror.CodeEngineEffortValueNotAllowed,
		"engine", engineName, "value", value, "allowed", strings.Join(allowed, ", "))
}

File: internal/ui/assets/js/validate.js
// 前端镜像的结构校验：对齐 internal/workflow/validate.go 的 ValidateStructured，供编辑器做即时
// 反馈（画布红点 / 顶栏「可运行」状态 / 拖拽连边即时拒环）。这不是终裁——服务端 ValidateStructured
// 才是终裁（保存 422 逐字段标错）；本文件须与 Go 版规则同步维护（见 ui.md〈已知限制〉）。
//
// 返回的 Problem 形状与服务端一致：{path, code, params, message}。code/params 是稳定身份，message
// 由当前 UI 字典渲染；path 用于定位字段，故本地问题与服务端 422 问题能共用同一套渲染逻辑。

import { NODE_ID_START, NODE_ID_END, isAgent, isMarker, detectCycle, ancestors } from "./graph.js";
import { formatProblem } from "./i18n.js";

export const NODE_ID_RE = /^[A-Za-z_][A-Za-z0-9_-]{0,63}$/;
const TEMPLATE_VAR_RE = /\\?\{\{([a-zA-Z_][\w.-]*)\}\}/g;
const KNOWN_SYS_VARS = new Set(["sys.userPrompt", "sys.cwd", "sys.runId"]);

// localProblems(def) → Problem[]。空定义 / 单条基础问题也走同一形状，调用方不必特判。
export function localProblems(def) {
  const nodes = def.nodes || [];
  const edges = def.edges || [];
  if (nodes.length === 0) {
    return [problem("nodes", "nodes_required")];
  }

  const problems = [];
  const indexByID = new Map();
  let startCount = 0,
    endCount = 0,
    agentCount = 0;
  nodes.forEach((node, i) => {
    if (node.id === NODE_ID_START) startCount++;
    else if (node.id === NODE_ID_END) endCount++;
    else agentCount++;
    if (!node.id) return; // 空 id 的必填错误在下方 agent 校验里报出
    if (indexByID.has(node.id)) {
      problems.push(problem(`nodes[${i}].id`, "duplicate_node_id", { id: node.id }));
      return;
    }
    indexByID.set(node.id, i);
  });
  const validNodeID = (id) => indexByID.has(id);

  if (startCount !== 1) problems.push(problem("nodes", "start_node_count", { count: startCount }));
  if (endCount !== 1) problems.push(problem("nodes", "end_node_count", { count: endCount }));
  if (agentCount === 0) problems.push(problem("nodes", "agent_node_required"));

  nodes.forEach((node, i) => {
    const path = `nodes[${i}]`;
    if (isMarker(node.id)) {
      if (node.displayName) problems.push(problem(path + ".displayName", "marker_field_not_empty", { id: node.id, field: "displayName" }));
      if (node.engine) problems.push(problem(path + ".engine", "marker_field_not_empty", { id: node.id, field: "engine" }));
      if (node.promptTemplate) problems.push(problem(path + ".promptTemplate", "marker_field_not_empty", { id: node.id, field: "promptTemplate" }));
      if (node.engineConfig) problems.push(problem(path + ".engineConfig", "marker_field_not_empty", { id: node.id, field: "engineConfig" }));
      return;
    }
    if (!node.id) problems.push(problem(path + ".id", "required_field", { field: "id" }));
    else if (!NODE_ID_RE.test(node.id)) problems.push(problem(path + ".id", "invalid_node_id", { id: node.id }));
    if (!node.displayName) problems.push(problem(path + ".displayName", "required_field", { field: "displayName" }));
    if (!node.engine) problems.push(problem(path + ".engine", "required_field", { field: "engine" }));
    if (!node.promptTemplate) problems.push(problem(path + ".promptTemplate", "required_field", { field: "promptTemplate" }));
    // engineConfig 的能力表校验（该引擎是否接受 model / effort 等）交服务端终裁：编辑器已按能力表
    // 条件渲染字段，正常操作路径不该产出非法组合，本地不重复这套判别联合。
  });

  const seen = new Set();
  edges.forEach((edge, i) => {
    const path = `edges[${i}]`;
    if (!edge.from || !edge.to) {
      problems.push(problem(path, "edge_endpoints_required"));
      return;
    }
    if (!validNodeID(edge.from)) problems.push(problem(path, "edge_from_node_not_found", { id: edge.from }));
    if (!validNodeID(edge.to)) problems.push(problem(path, "edge_to_node_not_found", { id: edge.to }));
    if (edge.from === edge.to) problems.push(problem(path, "self_edge", { from: edge.from, to: edge.to }));
    if (edge.from === NODE_ID_START && edge.to === NODE_ID_END) {
      problems.push(problem(path, "start_end_direct_edge"));
    } else {
      if (edge.to === NODE_ID_START) problems.push(problem(path, "edge_to_start"));
      if (edge.from === NODE_ID_END) problems.push(problem(path, "edge_from_end"));
    }
    const key = edge.from + "\u0000" + edge.to;
    if (seen.has(key)) problems.push(problem(path, "duplicate_edge", { from: edge.from, to: edge.to }));
    seen.add(key);
  });

  const inDeg = new Map(),
    outDeg = new Map();
  for (const e of edges) {
    outDeg.set(e.from, (outDeg.get(e.from) || 0) + 1);
    inDeg.set(e.to, (inDeg.get(e.to) || 0) + 1);
  }
  nodes.forEach((node, i) => {
    if (!isAgent(node.id)) return;
    const path = `nodes[${i}]`;
    if (!inDeg.get(node.id)) problems.push(problem(path, "node_missing_incoming_edge", { id: node.id }));
    if (!outDeg.get(node.id)) problems.push(problem(path, "node_missing_outgoing_edge", { id: node.id }));
  });

  const cycle = detectCycle(def);
  if (cycle) problems.push(problem("edges", "cycle_detected", { cycle: cycle.join("→") }));

  if (!cycle) {
    nodes.forEach((node, i) => {
      if (!isAgent(node.id)) return;
      const anc = ancestors(def, node.id);
      const path = `nodes[${i}].promptTemplate`;
      const text = node.promptTemplate || "";
      TEMPLATE_VAR_RE.lastIndex = 0;
      let m;
      while ((m = TEMPLATE_VAR_RE.exec(text))) {
        const [full, key] = m;
        if (full.startsWith("\\")) continue; // 转义 \{{x}} → 字面量，不校验
        if (key.startsWith("sys.")) {
          if (!KNOWN_SYS_VARS.has(key)) problems.push(problem(path, "unknown_system_variable", { key }));
          continue;
        }
        if (key === NODE_ID_START || key === NODE_ID_END) {
          problems.push(problem(path, "marker_node_reference", { id: key }));
          continue;
        }
        if (anc.has(key)) continue;
        if (validNodeID(key)) problems.push(problem(path, "non_ancestor_node_reference", { id: key }));
        else problems.push(problem(path, "node_reference_not_found", { id: key }));
      }
    });
  }

  return problems;
}

function problem(path, code, params) {
  const item = { path, code };
  if (params) item.params = params;
  item.message = formatProblem(item);
  return item;
}

// renameTemplateRef 把模板里的活引用 {{oldKey}} 改名为 {{newKey}}；转义的 \{{oldKey}} 是字面量、不动。
// 复用 TEMPLATE_VAR_RE 单一来源（与 render.go / validate.go 的 templateVariablePattern 同源），供编辑器改
// 节点 id 时级联同步下游模板引用。TEMPLATE_VAR_RE 带 /g，String.replace 会自行从头扫描并重置 lastIndex，
// 与本文件其它 exec 用法共享同一正则对象也互不干扰。
export function renameTemplateRef(template, oldKey, newKey) {
  return template.replace(TEMPLATE_VAR_RE, (full, key) => (!full.startsWith("\\") && key === oldKey ? `{{${newKey}}}` : full));
}

// nodeIdsWithProblems 把 Problem[] 里形如 nodes[i].xxx 的路径映射回节点 id 集合（供画布红点）。
export function nodeIdsWithProblems(def, problems) {
  const ids = new Set();
  for (const p of problems) {
    const m = /^nodes\[(\d+)\]/.exec(p.path);
    if (m && def.nodes[Number(m[1])]) ids.add(def.nodes[Number(m[1])].id);
  }
  return ids;
}
